hw/mcu/stm/stm32f4/Device/ST/STM32F4xx/Include/stm32f4xx.h

```diff
@@ -0,0 +1,269 @@
+/**

```

Comment:
It will be better if `hw/mcu/st` use this repo as submodule
https://github.com/micropython/stm32lib

It is maintained and updated by microphython, with all of the stm32 mcu family. Making porting tinyusb to circuitpython much easier later on.

Another minor thing is we should rename the vendor from stm to **st** which is more accurate.

Reply:
I agree w/ both changes, but I don't believe the repo you linked (https://github.com/micropython/stm32lib) corresponds to the header files I actually have.

I don't quite remember where I found the headers- @tannewt did the work for me there. Although I did _rediscover_ where he got them from when I started working on this.

Scott, could you point me to the repo or zip file where you found the source files under `hw/mcu/stm/stm32f4`?

Reply:
@tannewt do you have any suggestion on which repo is used for stm32lib, I am OK with anything that works :)

Reply:
@hathach I found the relevant zip file. The STM32F4-specific files that are _currently_ checked into tinyusb are _a subdirectory_ of the [Standard Peripheral Library](https://www.st.com/content/st_com/en/products/embedded-software/mcu-mpu-embedded-software/stm32-embedded-software/stm32-standard-peripheral-libraries/stsw-stm32065.html). Relative to the root of the zip file, the relevant files are under `STM32F4xx_DSP_StdPeriph_Lib_V1.8.0\Libraries\CMSIS`. The `DSP_Lib` directory from the zip file was omitted (it wasn't needed I imagine).

@ladyada, would it be okay if the Standard Peripheral Library was kept as a repository under the @adafruit account, as is [currently done](https://github.com/adafruit/asf4) with the Atmel headers?

If so, I think the entire repository should be kept and not just the subdirectory that made it into the tinyusb repo (@tannewt, feel free to disagree :P). Then I will convert `hw/st/smt32f4` into a submodule and update Makefile paths accordingly.

Reply:
seem like the std peripheral is legacy lib now, it is replaced by stm32cube lib, ST recommends that in @cr1901 's link. It is probably better to stick with latest lib for better support with new devices later on, unless it has a major drawback or restriction !!

Reply:
I have stm32cube installed; in general, stm32cube will ask you to _download_ all dependencies/headers from a GUI Wizard, and then place the all the dependencies for your current project inside your project directory. So projects are self-contained, but you need to generate the headers for each project using ST's IDE beforehand<sup>1</sup>.

The equivalent directory to the Standard Peripheral Library in an stm32cube project should be `Drivers\CMSIS`. Or at least it is on my machine :P.

1. Requiring the vendor IDE to get started seems par for the course for ARM vendors. NXP does it with lpcxpresso and ST seems to do it with stm32cube.

Reply:
@cr1901 It is troubleshome, sound similar to samd asf4 (must be a trend in silicon vendor :D), though it isresolvable. Micropython already download and put [stm32cube](https://github.com/micropython/stm32lib) library in the above link. We can either use it or folk it if needed.

Reply:
The repo you linked is missing the `Source` directory under `CMSIS/STM32F4xx`, which is needed to provide both the `.s` startup code and the `SystemInit()`/`SystemCoreClockUpdate()` functions in a separate `.c` file.

Additionally, as `board.mk` is currently coded, I've elected to use the CMSIS headers provided by the peripheral library (under `hw/mcu/stm/stm32f4/Include`) instead of the CMSIS headers you provided under `hw/cmsis`. Chances are, both copies are quite similar. But I should check that my code works with your copy (and see what the other MCU families do).

Reply:
I don't really remember where I got the files. I'd have do some sleuthing and don't have time for it at the moment.

Reply:
@cr1901 for the missing startup and linker script, we can add it in `hw/bsp`, I am not sure why micropython didn't include the system_stm32.c in the repo, maybe they want to customize it ( or it is how the stm32cube is designed to), but we can also store it in the repo in another folder under `hw/mcu/st`

Note; [system_stm32.c ](https://github.com/micropython/micropython/blob/master/ports/stm32/system_stm32.c) seems to be generic for the cube

@tannewt  no problems at all.

Reply:
for `cmsis` folder, yeah, we should use the one that provided by st in their library. Feel free to change is. `hw/cmsis` is included for mcu which the prph lib doesn't include cmsis such as Nordic's nrfx :)

Reply:
@hathach Okay, let's go with your idea, with one suggestion- can we store the startup code (`.s`) along with `system_stm32f4xx.c` files under `hw/mcu/st/startup/$family` and `hw/mcu/st/system-init` instead of under `hw/bsp`? The startup code is board agnostic, and none of the other TinyUSB ports seem to place startup code w/ the boards.

Just for completeness' sake, I'm gonna sum up what I've found (and maybe stick it into a README.md file):

The CMSIS by ARM is an attempt to unify how system initialization and headers for ARM microcontrollers are handled across vendors. A vendor is expected to provide a number of files, including vendor-specific device headers, vendor-specific system init code (in C), vendor-specific startup code (coded in assembly _and_ meant to make writing C interrupt handlers easy), and ARM architecture-specific headers. In practice, it's actually difficult to find all these files in one place.

* stm32cube generates a self-contained application. All the relevant headers, system init, and startup code are under `Drivers/CMSIS`. The linker script is stored at the root of your project, _along_ with a copy of the startup code. Relative to `Drivers/CMSIS`:
  * Device header files are stored under `Device/ST/$family/Include`.
  * CMSIS headers are stored under `Include`.
  * The portable C portion of the startup code, called `system_$family.c`, is stored under `Device/ST/$family/Source/Templates`. This "system init" code is required by CMSIS.
  * The assembly portion of the startup code, called `startup_$device.c`, is stored under `Device\ST\STM32F4xx\Source\Templates\gcc`.

* Micropython maintains a repo with _only_ the device header files. We use that repo and provide the remaining files ourselves.
  * Micropython keeps the linker script and assembly portion of the startup code under the [boards](https://github.com/micropython/micropython/tree/master/ports/stm32/boards) in the stm32 port (but the code is board agnostic).
  * Micropython also embeds its own [copy](https://github.com/micropython/micropython/tree/master/lib/cmsis/inc) of CMSIS headers for all ARM targets into the repo.
  * Micropython indeed _does_ use their own custom C startup code (CMSIS system init) for [stm32](https://github.com/micropython/micropython/blob/master/ports/stm32/system_stm32.c) targets. It is [missing](https://github.com/micropython/micropython/commit/fdbc22e4d27b9d1c686647c9c67a32206f7bc83a) the `SystemCoreClockUpdate` function.

* TinyUSB-specific info over where files are goes here- to be written :).

Reply:
@cr1901 thank you very much, yeah sure, you can put them in any folder under `hw/mcu/st`, the number of files and folder are rather small, it should be no issue at all. You could also move it around later on as well if needed, not a problem at all. I am looking forward to your next push and try it on my f4 discovery board here :D

Later on when integrating this into CPY, Adafruit may even wants to maintain Adafruit's own stm32lib like [Nordic nrfx](https://github.com/adafruit/nrfx), and [Samd asf4](https://github.com/adafruit/asf4) . It is only a matter of changing the submodule link reference :D

Reply:
@hathach I have updated the repo structure as requested. We use submodules when possible (stm32lib) and I have updated the remaining headers we keep around to be more up-to-date (from a dummy stm32cube project I used for testing).

I had to add a dummy `_init()` function to `hw/bsp/stm32f407g_disc1/board_stm32f407g_disc1.c` to satisfy the linker (this seems to be a known issue- CTRL+F "_init" on [this page](https://github.com/bjornfor/stm32-test)). And of course I had to update some paths. But other than that, the demo still works fine. I even collected my thoughts into `hw/mcu/st/README.md` so I don't forget the layout :P.

@tannewt, @ladyada, @ptorrone: If @hathach approves my changes now, is it okay if we merge and deal with further issues/fixes in subsequent PRs and commits to keep momentum going?